File: Benchmarks/Core/Types/Vector2.cs
using System.Runtime.CompilerServices;

namespace Benchmarks.Core.Types
{
    public struct Vector2(float x, float y)
    {
        public float X { get; set; } = x;
        public float Y { get; set; } = y;

        public static readonly Vector2 Zero = new(0, 0);
        public static readonly Vector2 One = new(1, 1);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector2 operator +(Vector2 a, Vector2 b) => new(a.X + b.X, a.Y + b.Y);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector2 operator -(Vector2 a, Vector2 b) => new(a.X - b.X, a.Y - b.Y);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector2 operator *(Vector2 a, float s) => new(a.X * s, a.Y * s);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool operator ==(Vector2 a, Vector2 b) => a.X == b.X && a.Y == b.Y;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool operator !=(Vector2 a, Vector2 b) => !(a == b);

        public override bool Equals(object obj) => obj is Vector2 v && this == v;
        public override int GetHashCode() => HashCode.Combine(X, Y);
    }
}

File: Benchmarks_FSharp/Core/Types/Vector2.fs
namespace Benchmarks.Core.Types

open System.Runtime.CompilerServices

[<Struct>]
type Vector2 =
    val mutable X: float32
    val mutable Y: float32
    new(x, y) = { X = x; Y = y }

    static member Zero = Vector2(0f, 0f)
    static member One = Vector2(1f, 1f)

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    static member (+)(a: Vector2, b: Vector2) = Vector2(a.X + b.X, a.Y + b.Y)

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    static member (-)(a: Vector2, b: Vector2) = Vector2(a.X - b.X, a.Y - b.Y)

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    static member (*)(a: Vector2, s: float32) = Vector2(a.X * s, a.Y * s)

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    static member op_Equality(a: Vector2, b: Vector2) = a.X = b.X && a.Y = b.Y

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    static member op_Inequality(a: Vector2, b: Vector2) = not (Vector2.op_Equality(a, b))

    override this.Equals(obj) =
        match obj with
        | :? Vector2 as v -> this.X = v.X && this.Y = v.Y
        | _ -> false

    override this.GetHashCode() = HashCode.Combine(this.X, this.Y)
